### ISSUE TITLE
Implement a efficient way to find the number of files in a dataset

### ISSUE THREAD

**Issue by vijayphoenix** [tensorflow/datasets#1403]
**Is your feature request related to a problem? Please describe.**
We can obtain the no. of files in a dataset in the following way.
```
_, info = tfds.load('cats_vs_dogs', with_info=True)
print(info.splits['train'].num_examples)
```
But this solution does not work when we define our own train and test split. For example:
```
split = tfds.Split.TRAIN.subsplit([90, 10])
(training_set, validation_set), info = tfds.load('cats_vs_dogs', with_info=True, as_supervised=True, split=split)
```
Here we cannot use the above method for finding the length of `training_set`.

**Describe the solution you'd like**
It would be great if we could have a function or variable through which we can obtain the no. of files in the dataset. 

**Describe alternatives you've considered**
Some inefficient solutions I could think of are

```
num_training_examples = 0
for _ in training_set:
    num_training_examples += 1

or 

len(list(training_set))
```

**Comment by Conchylicultor** [comment]
Yes, it should be possible nice to expose this information. For the API, how about something like:
How about allowing something like ?

```
info.splits['train[50%:]'].num_examples
```

It should be relatively straightforward to wrap it in a function and expose it to the public API.
This can be done in two steps:

1. Add a `_num_examples_from_split_dict(info.splits, 'train[50%:]')` util function
1. Modify `SplitDict.__getitem__` to expose this function

## Instructions for 1

Currently, this is done in `_read_single_instruction(`:
https://github.com/tensorflow/datasets/blob/master/tensorflow_datasets/core/tfrecords_reader.py

```
  if not isinstance(instruction, ReadInstruction):
    instruction = ReadInstruction.from_spec(instruction)
  absolute_instructions = instruction.to_absolute(name2len)
```
with:
```
name2shard_lengths = {info.name: info.shard_lengths for info in split_infos}
name2len = {
    name: sum(lengths)
    for name, lengths in name2shard_lengths.items()
}
```

## Instructions for 2

This would require modifying the `SplitDict.__getitem__` class to call and return a dummy `DummySplitInfo(num_examples=_num_examples_from_split_dict(...))` when called with a non-split. https://github.com/tensorflow/datasets/blob/2b1f6051a560b1d0dd2cfb8c31bb778d7d290a6c/tensorflow_datasets/core/splits.py#L533

We won't have time for this in the short term. Don't hesitate if anyone want to take over this issue. I'm happy to help more if you get stuck.

**Comment by vijayphoenix** [comment]
@Conchylicultor Thank you for your input.
You can assign me to this issue, I will be happy to help.

**Comment by Conchylicultor** [comment]
Thank you very much, this would be great.

**Comment by Conchylicultor** [comment]
I ended up taking this as this was required by other people. It should appear in tfds-nightly within a few days.

**Comment by vijayphoenix** [comment]
Well, sorry for the delay on my side and thanks for the API update

**Comment by Conchylicultor** [comment]
No worries, thank you for the help. It ended up to be a little more tricky than I anticipated.

**Comment by Flamefire** [comment]
I was wondering if there was a better way to do this. I noticed that tensorflow during the first epoch shows the training progress as something like `32/Unknown` while it shows the correct number instead of "Unknown" during the 2nd epoch.

So there seems to be something that knows the size. It would be great if we could just do `dataset.num_examples` or even `len(dataset)` (overloaded `__len__`) which even survives the wrappers like `map` and `shuffle`.

This is especially useful for using distrtibuted training because `MultiWorkerMirroredStrategy` needs to know the number of samples/batches: https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras#train_the_model_with_multiworkermirroredstrategy
> Since MultiWorkerMirroredStrategy does not support last partial batch handling, pass the steps_per_epoch argument to model.fit() when dataset is imbalanced on multiple workers.

**Comment by Conchylicultor** [comment]
@Flamefire TFDS is a wrapper around a `tf.data.Dataset` object. There is currently no way to specify the `len` of the dataset in `tf.data.Dataset` as far as I know.

That could be a nice feature to add to TensorFlow though, but it's not something TFDS can do anything about it while TF do not supports `__len__`.

**Comment by Flamefire** [comment]
There is something similar: `tensorflow.python.data.experimental.ops.cardinality`

Example:

    import tensorflow as tf
    from tensorflow.python.data.experimental.ops import cardinality
    import numpy as np

    scale = lambda x: x
    BATCH_SIZE = 64
    examples = np.ones([10*BATCH_SIZE,28,28,1])
    labels = np.ones([examples.shape[0]])
    dataset = tf.data.Dataset.from_tensor_slices((examples, labels))
    x = dataset.map(scale).cache().shuffle(10000).batch(BATCH_SIZE)
    card = cardinality.cardinality(x)

This returns `10`.
However for TFDS this does not work and returns "Unknown"(-2).

From some searching in the TF source I found a function `DatasetBase::Cardinality` which is implemented for a lot of Datasets and DatasetOps which may modify the Cardinality based on the input (e.g. the `batch` does basically: `return input->Cardinality() / batch_size`)

Maybe you could check at which point this breaks for TFDS. Either TFDS is trying to be to clever or there is something missing in TF.

BTW: Why are you not simply returning a TFRecordDataset (included in TF) and let TF handle the reading and other stuff?

**Comment by Conchylicultor** [comment]
> Why are you not simply returning a TFRecordDataset (included in TF) and let TF handle the reading and other stuff?

We are returning `TFRecordDataset`, but as you pointed out, `TFRecordDataset` do not provides cardinality.

```
ds = tf.data.TFRecordDataset()
tf.data.experimental.cardinality(ds) == tf.data.experimental.UNKNOWN_CARDINALITY
```

If there was a way to manually overwrite the cardinality, we could provide this information to `tf.data.Dataset` object as we can compute the number of example after the interleave op. But that would be a feature to add to core TF.

**Comment by Conchylicultor** [comment]
I'm closing this as the original issue has been solved. I opened https://github.com/tensorflow/datasets/issues/1456 to track the cardinality issue.

**Comment by Pavan0798** [comment]
can anyone please let me know.  info.splits['train[50%:]'].**num_examples****** what is num_examples here
